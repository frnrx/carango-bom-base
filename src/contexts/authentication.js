import React, { useMemo } from 'react';

import useAuth from '../hooks/useAuth';

export const AuthenticationContext = React.createContext({});

const AuthenticationProvider = ({ children }) => {
  const authState = useAuth();
  const value = useMemo(
    () => ({
      isLoggedIn: authState.isLoggedIn,
      isLoading: authState.isLoading,
      login: authState.login,
      logout: authState.logout,
    }),
    [authState],
  );

  return (
    <AuthenticationContext.Provider value={value}>
      {children}
    </AuthenticationContext.Provider>
  );
}

export default AuthenticationProvider;
