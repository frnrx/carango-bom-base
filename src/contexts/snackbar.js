import React, { createContext, useState, useMemo, useCallback } from 'react';

import Snackbar from '@material-ui/core/Snackbar';
import Alert from '@material-ui/lab/Alert';

export const SnackBarContext = createContext();

const SnackBarProvider = ({ children }) => {
  const [open, setOpen] = useState(false);
  const [alert, setAlert] = useState();
  const [severity, setSeverity] = useState('info');
  const [duration, setDuration] = useState(6000);

  const handleClose = () => {
    setOpen(false);
  };

  const addAlert = useCallback(({ content, customSeverity, customDuration }) => {
    setOpen(true);
    setAlert(content);
    if (customSeverity) setSeverity(customSeverity);
    if (customDuration) setDuration(customDuration);
  }, []);

  const value = useMemo(
    () => ({
      addAlert,
    }),
    [addAlert],
  );

  return (
    <SnackBarContext.Provider value={value}>
      <>
        {children}
        <Snackbar
          open={open}
          autoHideDuration={duration}
          onClose={handleClose}
          anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
        >
          <Alert severity={severity}>{alert}</Alert>
        </Snackbar>
      </>
    </SnackBarContext.Provider>
  );
};

export default SnackBarProvider;
