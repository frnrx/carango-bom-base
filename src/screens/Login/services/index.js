import { API_URL } from "../../../services/constants";
import fetchResponseHandler from "../../../services/fetchResponseHandler";

export const loginService = async (email, password) => {
  const requestOptions = {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ email, senha: password }),
  };
  return fetch(`${API_URL}/auth`, requestOptions).then(fetchResponseHandler);
};
