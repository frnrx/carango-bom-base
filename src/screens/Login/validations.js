const validations = {
  email: (data) => {
    const emailRegex = new RegExp(/\S+@\S+\.\S+/);
    if (data && emailRegex.test(data)) {
      return { valid: true, showError: false };
    }
    return { valid: false, text: 'E-mail invÃ¡lido.', showError: true };
  },
  password: (data) => {
    if (data && data.length >= 6) {
      return { valid: true, showError: false };
    }
    return { valid: false, text: 'Senha deve ter ao menos 6 caracteres.', showError: true };
  },
};

export default validations;
