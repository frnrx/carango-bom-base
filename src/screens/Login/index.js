import React, { useState, useContext, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { TextField, Grid, CircularProgress } from '@material-ui/core';

import useErrors from '../../hooks/useErrors';
import { AuthenticationContext } from '../../contexts/authentication';

import FormButton from '../../components/FormButton';

import validations from './validations';

const Login = () => {
  const history = useHistory();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const { isLoggedIn, login, isLoading } = useContext(AuthenticationContext);

  const [errors, validateFields, isFormValid] = useErrors(validations);

  const onSubmit = (event) => {
    event.preventDefault();

    login(email, password);
  };

  const handleUpdateEmail = (event) => {
    setEmail(event.target.value);
  };

  const handleUpdatePassword = (event) => {
    setPassword(event.target.value);
  };
  useEffect(() => {
    if (isLoggedIn) {
      history.push('/');
    }
  }, [isLoggedIn]);

  return (
    <form onSubmit={onSubmit}>
      <Grid container direction="column" spacing={2} alignContent="center">
        <Grid container item xs={8}>
          <TextField
            fullWidth
            id="email"
            label="E-mail"
            type="email"
            name="email"
            role="textbox"
            onChange={handleUpdateEmail}
            variant="outlined"
            onBlur={validateFields}
            helperText={errors.email.text}
            error={errors.email.showError}
          />
        </Grid>
        <Grid container item xs={8}>
          <TextField
            fullWidth
            id="password"
            label="Senha"
            type="password"
            name="password"
            role="textbox"
            onChange={handleUpdatePassword}
            variant="outlined"
            onBlur={validateFields}
            helperText={errors.password.text}
            error={errors.password.showError}
          />
        </Grid>
        <Grid item container justify="space-between" xs={8}>
          <FormButton to="/" isLink>
            Voltar
          </FormButton>
          <FormButton
            color="primary"
            type="submit"
            disabled={!isFormValid()}
          >
            {isLoading ? <CircularProgress /> : 'Logar'}
          </FormButton>
        </Grid>
      </Grid>
    </form>
  );
};

export default Login;
