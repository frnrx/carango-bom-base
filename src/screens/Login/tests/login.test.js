import React from 'react';
import { render, screen, fireEvent, cleanup } from '@testing-library/react';

import Login from '..';

import { AuthenticationContext } from '../../../contexts/authentication';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe('Login screen', () => {
  it('should match the snapshot', () => {
    const { container } = render(<Login />);

    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('Login form', () => {
  let emailInput;
  let passwordInput;
  let submitButton;
  let goBackButton;

  beforeEach(() => {
    render(<Login />);

    emailInput = screen.getByRole('textbox', { name: /E-mail/i });
    passwordInput = screen.getByLabelText(/Senha/i);
    submitButton = screen.getByRole('button', { name: /Logar/i });
    goBackButton = screen.getByRole('button', { name: /Voltar/i });
  });

  it('should render the login form', () => {
    expect(emailInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });

  it('should send the data correctly when a user presses submit', () => {
    cleanup();

    const mockedState = {
      isLoggedIn: false,
      login: jest.fn(),
    };

    const { getByRole, getByLabelText } = render(
      <AuthenticationContext.Provider value={mockedState}>
        <Login />
      </AuthenticationContext.Provider>,
    );

    const emailInputWithContext = getByRole('textbox', { name: /E-mail/i });
    const passwordInputWithContext = getByLabelText(/Senha/i);
    const submitButtonWithContext = getByRole('button', { name: /Logar/i });

    fireEvent.change(emailInputWithContext, {
      target: {
        value: 'johndoe@doe.com',
      },
    });

    fireEvent.focusOut(emailInputWithContext);

    fireEvent.change(passwordInputWithContext, {
      target: {
        value: '123456',
      },
    });

    fireEvent.focusOut(passwordInputWithContext);

    expect(submitButtonWithContext).not.toBeDisabled();

    fireEvent.click(submitButtonWithContext);

    expect(mockedState.login).toHaveBeenCalledWith('johndoe@doe.com', '123456');
  });

  it('should only enable the submit button when all fields are filled', () => {
    expect(submitButton).toHaveAttribute('disabled');

    fireEvent.change(emailInput, {
      target: {
        value: 'johndoe@joe.com',
      },
    });

    fireEvent.focusOut(emailInput);

    expect(submitButton).toHaveAttribute('disabled');

    fireEvent.change(passwordInput, {
      target: {
        value: '123456789',
      },
    });

    fireEvent.focusOut(passwordInput);

    expect(submitButton).not.toBeDisabled();
  });

  it('should keep the button disabled when the email don`t meet the requirements', () => {
    fireEvent.change(emailInput, {
      target: {
        value: 'joe',
      },
    });

    fireEvent.focusOut(emailInput);

    const errorText = screen.getByText('E-mail invÃ¡lido.');

    expect(errorText).toBeInTheDocument();

    expect(submitButton).toHaveAttribute('disabled');
  });

  it('should keep the button disabled when the password don`t meet the requirements', () => {
    fireEvent.change(passwordInput, {
      target: {
        value: '123',
      },
    });

    fireEvent.focusOut(passwordInput);

    const errorText = screen.getByText('Senha deve ter ao menos 6 caracteres.');

    expect(errorText).toBeInTheDocument();

    expect(submitButton).toHaveAttribute('disabled');
  });

  it('should should go to home if the user presses Voltar', () => {
    fireEvent.click(goBackButton);

    expect(mockHistoryPush).toHaveBeenCalledWith('/');
  });
});

describe('login authorization provider usage', () => {
  it('should redirect the user to the home if they are already logged', () => {
    const mockedState = {
      isLoggedIn: true,
    };
    render(
      <AuthenticationContext.Provider value={mockedState}>
        <Login />
      </AuthenticationContext.Provider>,
    );

    expect(mockHistoryPush).toHaveBeenCalledWith('/');
  });

  it('should not redirect the user to the home if they are not logged', () => {
    const mockedState = {
      isLoggedIn: false,
    };
    render(
      <AuthenticationContext.Provider value={mockedState}>
        <Login />
      </AuthenticationContext.Provider>,
    );

    expect(mockHistoryPush).not.toHaveBeenCalledWith('/');
  });

  it('should find the loading component if the login is still loading', () => {
    const mockedState = {
      isLoggedIn: false,
      isLoading: true,
    };
    render(
      <AuthenticationContext.Provider value={mockedState}>
        <Login />
      </AuthenticationContext.Provider>,
    );

    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });
});
