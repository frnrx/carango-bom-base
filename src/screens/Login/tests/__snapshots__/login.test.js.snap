// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen should match the snapshot 1`] = `
<form>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-content-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-8"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        role="textbox"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="email"
          id="email-label"
        >
          E-mail
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="email"
            name="email"
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                E-mail
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-8"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        role="textbox"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Senha
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Senha
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-space-between MuiGrid-grid-xs-8"
    >
      <div
        class="MuiGrid-root makeStyles-buttonContainer-5 MuiGrid-item"
      >
        <a
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-6 MuiButton-fullWidth"
          href="/"
          role="link"
          tabindex="0"
        >
          <span
            class="MuiButton-label"
          >
            Voltar
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </div>
      <div
        class="MuiGrid-root makeStyles-buttonContainer-5 MuiGrid-item"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-6 MuiButton-containedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled"
          disabled=""
          role="button"
          tabindex="-1"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Logar
          </span>
        </button>
      </div>
    </div>
  </div>
</form>
`;
